File: BabylonCS.JS/lib/engine.ts
import { Engine } from "@babylonjs/core";

export const createEngine = (
  canvas: HTMLCanvasElement,
  antialias: boolean
): Engine => new Engine(canvas, antialias);

export const engineRunRenderLoop = (engine: Engine, callback: () => void) =>
  engine.runRenderLoop(callback);

export const engineResize = (engine: Engine) => engine.resize();

export const engineSetupResize = (engine: Engine) => {
  window.addEventListener("resize", () => engineResize(engine));
};

export const engineModule = {
  createEngine,
  engineRunRenderLoop,
  engineResize,
  engineSetupResize,
};

File: BabylonCS/BabylonCS/Engine.cs
using System.Runtime.InteropServices.JavaScript;
using BabylonCS.Interop;

namespace BabylonCS;

public interface IEngine
{
    void RunRenderLoop(Action callback);
    void SetupResize();
}

public class Engine(RenderCanvas renderCanvas, bool antialias) : IEngine
{
    internal JSObject EngineObj { get; } = EngineImpl.CreateEngine(renderCanvas.CanvasElement, antialias);

    public void RunRenderLoop(Action callback)
    {
        EngineImpl.EngineRunRenderLoop(EngineObj, callback);
    }

    public void SetupResize()
    {
        EngineImpl.EngineSetupResize(EngineObj);
    }
}